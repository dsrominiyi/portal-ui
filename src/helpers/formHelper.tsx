import React, { ReactNode } from 'react';

import QuestionBox from '../components/QuestionBox';
import TextField, { Props as TextFieldProps } from '../components/TextField';
import AbiSearch, { Props as AbiSearchProps } from '../components/AbiSearch';
import Dropdown, { DropdownOption, Props as DropdownProps } from '../components/Dropdown';
import MultiChoice, { Props as MultiChoiceProps } from '../components/MultiChoice';
import DateFields, { Props as DateFieldsProps } from '../forms/common/DateFields';
import Lookup from '../components/Lookup';

import { ValidationError } from './formValidator';

import formStyles from '../forms/styles.scss';

interface InputConfig {
  type: InputType;
  label: string;
  buttonLabel?: string;
  secondaryButtonLabel?: string;
  value?: FieldValue;
  codeListId?: string;
  options?: DropdownOption[] | string[];
  attributes?: { [attr: string]: string };
  onChange?:
    | TextFieldProps['onChange']
    | AbiSearchProps['onChange']
    | DropdownProps['onChange']
    | DateFieldsProps['onChange']
    | MultiChoiceProps['onClick'];
  secondaryAction?: () => void;
}

export const buildInput = (config: InputConfig, fieldError?: ValidationError) => {
  let input: ReactNode;

  switch (config.type) {
    case InputType.Text:
      input = (
        <TextField
          key="0"
          additionalClasses={[formStyles.formTextField]}
          label={config.label}
          value={config.value as string}
          isInvalid={fieldError?.showError}
          onChange={config.onChange as TextFieldProps['onChange']}
          attributes={config.attributes}
        />
      );
      break;
    case InputType.Lookup:
      input = (
        <Lookup
          key="0"
          label={config.label}
          buttonLabel={config.buttonLabel}
          secondaryButtonLabel={config.secondaryButtonLabel}
          value={config.value as string}
          isInvalid={fieldError?.showError}
          onChange={config.onChange as TextFieldProps['onChange']}
          secondaryAction={config.secondaryAction}
          attributes={config.attributes}
        />
      );
      break;
    case InputType.AbiSearch:
      input = (
        <AbiSearch
          key="0"
          label={config.label}
          codeListId={config.codeListId || ''}
          onChange={config.onChange as AbiSearchProps['onChange']}
          attributes={config.attributes}
        />
      );
      break;
    case InputType.Dropdown:
      input = (
        <Dropdown
          key="0"
          label={config.label}
          options={config.options!}
          selected={config.value as string}
          isInvalid={fieldError?.showError}
          onChange={config.onChange as DropdownProps['onChange']}
          attributes={config.attributes}
        />
      );
      break;
    case InputType.DateFields:
      input = (
        <DateFields
          key="0"
          values={config.value as DateFieldValues}
          onChange={config.onChange as DateFieldsProps['onChange']}
          testId={config.attributes && config.attributes['test-id']}
        />
      );
      break;
    case InputType.MultiChoice:
      input = (
        <MultiChoice
          key="0"
          options={config.options as string[]}
          selectedOption={config.value as string}
          onClick={config.onChange as MultiChoiceProps['onClick']}
          attributes={config.attributes}
        />
      );
      break;
  }

  return (
    <QuestionBox
      key={config.label}
      text={[InputType.DateFields, InputType.MultiChoice].includes(config.type) ? config.label : ''}
      hasError={fieldError?.showError}
      errorText={fieldError?.message}
      inputFields={[input]}
    />
  );
};
