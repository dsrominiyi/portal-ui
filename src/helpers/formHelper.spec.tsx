import { shallow } from 'enzyme';

import { buildInput } from './formHelper';

describe('formHelper', () => {
  describe('buldInput', () => {
    test('AbiSearch input passes and empty string as the codeListId if none is provided', () => {
      const inputBox = buildInput({
        type: InputType.AbiSearch,
        label: 'test',
        onChange: () => null
      });

      expect(inputBox.props.inputFields[0].props.codeListId).toBe('');
    });

    test('Dropdown input with validation errors matches snapshot', () => {
      const inputBox = buildInput(
        {
          type: InputType.Dropdown,
          label: 'test',
          options: ['1', '2'],
          value: '1',
          onChange: () => null
        },
        { showError: true, message: 'invalid' }
      );

      expect(shallow(inputBox)).toMatchSnapshot();
    });
  });
});
