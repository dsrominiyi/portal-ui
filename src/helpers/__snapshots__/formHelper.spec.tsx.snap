// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formHelper buldInput Dropdown input with validation errors matches snapshot 1`] = `
<div
  className="questionBox hasError"
>
  <div>
    <Dropdown
      isInvalid={true}
      key="0"
      label="test"
      onChange={[Function]}
      options={
        Array [
          "1",
          "2",
        ]
      }
      selected="1"
    />
  </div>
  <div
    className="errorText"
  >
    invalid
  </div>
</div>
`;
